src/OrbitQt/HistogramWidget.cpp

```diff
@@ -87,74 +93,129 @@ static void DrawVerticalLine(QPainter& painter, const QPoint& start, int length)
+                               double max_value) {
   DrawHorizontalLine(painter, axes_intersection, length);
 
-  const int tick_spacing_pixels =
-      RoundToClosestInt(static_cast<double>(length) / kHorizontalTickCount);
+  const QFontMetrics font_metrics(painter.font());
 
-  int current_tick_location = axes_intersection.x();
-  uint64_t current_tick_value = min_value;
+  for (size_t i = 0; i < ticks.values.size(); ++i) {
+    const double tick_value = ticks.values[i];
+    const int tick_location =
+        ValueToAxisLocation(tick_value, axis_length, min_value, max_value) + axes_intersection.x();
 
-  const QFontMetrics font_metrics(painter.font());
+    if (!(axes_intersection.x() <= tick_location &&
+          tick_location <= axes_intersection.x() + axis_length)) {
+      continue;
+    }
 
-  for (uint32_t i = 0; i <= kHorizontalTickCount; ++i) {
-    DrawVerticalLine(painter, {current_tick_location, axes_intersection.y()},
-                     kHorizontalAxisTickLength);
+    DrawVerticalLine(painter, {tick_location, axes_intersection.y()}, kHorizontalAxisTickLength);
 
-    const QString tick_label =
-        QString::number(orbit_display_formats::ToDoubleInGivenTimeUnits(
-                            absl::Nanoseconds(current_tick_value), time_unit),
-                        'f', kTimeDecimalsCount);
+    const QString& tick_label = ticks.labels[i];
     const QRect tick_label_bounding_rect = font_metrics.tightBoundingRect(tick_label);
 
-    painter.drawText(current_tick_location - tick_label_bounding_rect.width() / 2,
+    painter.drawText(tick_location - tick_label_bounding_rect.width() / 2,
                      axes_intersection.y() + kHorizontalAxisTickLength +
                          tick_label_bounding_rect.height() + kTickLabelGap + kLineWidth / 2,
                      tick_label);
-
-    current_tick_location += tick_spacing_pixels;
-    current_tick_value += tick_spacing_as_value;
   }
 }
 
 static void DrawVerticalAxis(QPainter& painter, const QPoint& axes_intersection, int length,
-                             double max_freq) {
+                             const Ticks& ticks, int axis_length, double max_value) {
   DrawVerticalLine(painter, axes_intersection, -length);
 
-  const double tick_spacing_as_value = max_freq / kVerticalTickCount;
-  const int tick_spacing_pixels =
-      RoundToClosestInt(static_cast<double>(length) / kVerticalTickCount);
+  const QFontMetrics font_metrics(painter.font());
 
-  double current_tick_value = tick_spacing_as_value;
-  int current_tick_location = axes_intersection.y() - tick_spacing_pixels;
+  for (size_t i = 1; i < ticks.values.size(); ++i) {
+    const double tick_value = ticks.values[i];
+    const int tick_location =
+        axes_intersection.y() - ValueToAxisLocation(tick_value, axis_length, 0.0, max_value);
 
-  const QFontMetrics font_metrics(painter.font());
+    if (!(axes_intersection.y() - axis_length <= tick_location &&
```

Comment:
Nit: Maybe adding a comment before, "skipping ticks outside the visible axis range" or something like that.